### PR TITLE
feat: borrow and repay assets

### LINKED ISSUE #92: x/leverage: Support Borrowing Assets
## Summary

A user should be able to borrow allow-list assets via the `x/leverage` module using "u" token collateral.

Note, without an oracle in place, all integration tests should be done with borrowing and lending with internal tokens.

----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by toteki**
## Description

Introduces `MsgSetCollateral`, `MsgBorrowAsset`, `MsgRepayAsset` and all supporting handlers, errors, and keeper keys.

PR does not include events, logging, or testing.

Uses placeholder values in place of Oracle integration.

Closes #92

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by toteki** [comment]
I still have major TODOs in keeper.go to cover, but the rest of the PR seems tidy so far.

I was hoping to create a second PR after this one merges, to cover events, logging, and testing associated with the features here.

If the current PR stays tidy (and the above PR follows shortly after), it can become a good model of "here are the files to change when adding a new message type" for my future self / others.

File: go.sum [hunk 1]
```diff
@@ -1088,7 +1088,6 @@ github.com/jgautheron/goconst v1.5.1 h1:HxVbL1MhydKs8R8n/HE5NPvzfaYmQJA3o879lE4+
 github.com/jgautheron/goconst v1.5.1/go.mod h1:aAosetZ5zaeC/2EfMeRswtxUFBpe2Hr7HzkgX4fanO4=
 github.com/jhump/protoreflect v1.6.1/go.mod h1:RZQ/lnuN+zqeRVpQigTwO6o0AJUkxbnSnpuG7toUTG4=
 github.com/jhump/protoreflect v1.8.2/go.mod h1:7GcYQDdMU/O/BBrl/cX6PNHpXh6cenjd8pneu5yW7Tg=
```
Comment (toteki) [inline go.sum:1090]:
```suggestion
github.com/jhump/protoreflect v1.8.2/go.mod h1:7GcYQDdMU/O/BBrl/cX6PNHpXh6cenjd8pneu5yW7Tg=
github.com/jhump/protoreflect v1.9.0 h1:npqHz788dryJiR/l6K/RUQAyh2SwV91+d1dnh4RjO9w=
```

File: x/leverage/types/keys.go [hunk 2]
```diff
@@ -46,3 +52,47 @@ func CreateRegisteredTokenKey(baseTokenDenom string) []byte {
...
+	//	e.g. if KeyPrefixLoanToken | lengthPrefixed(borrowerAddr.String()) were resolved to
+	//		0x04 | 0x03 0x41 0x42 0x43 (example address string simplified to "ABC")
+	//	then the iterator start/end would be
+	//		0x04 | 0x03 0x41 0x42 0x43 (inclusive start)
+	//		0x04 | 0x03 0x41 0x42 0x44 (exclusive end)
+	//	and keys like the following would fall within the range
+	//		0x04 | 0x03 0x41 0x42 0x43 | ... (any key that has prefix)
+	//
+	//	I couldn't find documentation on this behavior but it seems like
+	//		how it would reasonably work if sdk.Iterators are prefix-friendly.
```
Comment (toteki) [inline x/leverage/types/keys.go:81]:
```suggestion
	//	Question: Is this the right way to derive a range for an sdk.Iterator?
	//
	//	e.g. if KeyPrefixLoanToken | lengthPrefixed(borrowerAddr.String()) were resolved to
	//		0x04 | 0x03 0x41 0x42 0x43 (example address string simplified to "ABC")
	//	then the iterator start/end would be
	//		0x04 | 0x03 0x41 0x42 0x43 (inclusive start)
	//		0x04 | 0x03 0x41 0x42 0x44 (exclusive end)
	//	and keys like the following would fall within the range
	//		0x04 | 0x03 0x41 0x42 0x43 | ... (any key that has prefix)
	//
	//	I couldn't find documentation on this behavior but it seems like
	//		how it would reasonably work if sdk.Iterators are prefix-friendly.
```
Note: Question

File: x/leverage/keeper/keeper.go [hunk 3]
```diff
@@ -123,3 +125,157 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
+func (k Keeper) GetAllLoans(ctx sdk.Context, borrowerAddr sdk.AccAddress) (sdk.Coins, error) {
+	currentlyBorrowed := sdk.NewCoins()
+	store := ctx.KVStore(k.storeKey)
+	var key []byte
+	key = append(key, types.KeyPrefixLoanToken...)
+	key = append(key, address.MustLengthPrefix(borrowerAddr)...)
+	prefixStore := prefix.NewStore(store, key)
+	iter := prefixStore.Iterator(nil, nil) // TODO: Find out why iter starts with valid = false
+	//	Alternative:
+	//  iter := sdk.KVStorePrefixIterator(store, key) // TODO: Same problem
```
Comment (toteki) [inline x/leverage/keeper/keeper.go:165]: Note: This is the place where iterator is invalid when created. It might also be a problem with the prefixStore or the key.

Comment (toteki) [reply x/leverage/keeper/keeper.go]: Note to self: it was a misunderatanding on my part - it's _supposed_ to be invalid once all prefixed keys have been iterated through (including when created and no qualifying keys exist).

File: x/leverage/types/keys.go [hunk 4]
```diff
@@ -41,8 +50,30 @@ func CreateUTokenDenomKey(uTokenDenom string) []byte {
...
 func CreateRegisteredTokenKey(baseTokenDenom string) []byte {
+	// assetprefix | denom | 0x00
 	var key []byte
 	key = append(key, KeyPrefixRegisteredToken...)
 	key = append(key, []byte(baseTokenDenom)...)
 	return append(key, 0) // append 0 for null-termination
 }
+
+// CreateLoanKey returns a KVStore key for getting and setting a Loan in a single denom and borrower address
+func CreateLoanKey(borrowerAddr sdk.AccAddress, tokenDenom string) []byte {
```
Comment (alexanderbez) [inline x/leverage/types/keys.go:61]: For this loan key, we're storing the `sdk.Int` amount of the loan, correct?

File: x/leverage/keeper/keeper_test.go [hunk 5]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
 	suite.Require().Equal(int64(0), uTokenBalance.Amount.Int64())
 }
 
+// Helper function: Initialize the common starting scenario from which borrow and repay tests stem:
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:150]:
```suggestion
// initBorrowScenario initializes ...
```

File: x/leverage/keeper/keeper_test.go [hunk 6]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
...
+	// lender lends 1k umee and receives 1k u/umee
+	err := app.LeverageKeeper.LendAsset(ctx, lenderAddr, tCoin("umee", 1000))
+	suite.Require().NoError(err)
+	// return the three account addresses
+	return lenderAddr, bumAddr
+	// The starting scenario is thus:
+	// - umee and u/umee are accepted assets
+	// - a "lender" user has 9k umee and 1k u/umee
+	// - the leverage module has 1k umee due to lender's lending
+	// - a "bum" user has an address but no assets
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:175]: Do we need this comment? Is there more to do in this function? Can we move it to the godoc?

File: x/leverage/keeper/keeper_test.go [hunk 7]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
 	suite.Require().Equal(int64(0), uTokenBalance.Amount.Int64())
 }
 
+// Helper function: Initialize the common starting scenario from which borrow and repay tests stem:
+func (suite *IntegrationTestSuite) initBorrowScenario() (lender, bum sdk.AccAddress) {
+	app, ctx := suite.app, suite.ctx
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:152]: nit: try to add newlines after logical groupings of calls. This method is way too difficult to read.

File: x/leverage/keeper/keeper_test.go [hunk 8]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
...
+	// The starting scenario is thus:
+	// - umee and u/umee are accepted assets
+	// - a "lender" user has 9k umee and 1k u/umee
+	// - the leverage module has 1k umee due to lender's lending
+	// - a "bum" user has an address but no assets
+}
+
+func (suite *IntegrationTestSuite) TestBorrowAsset_Invalid() {
+	lenderAddr, _ := suite.initBorrowScenario() // create initial conditions
+	app, ctx := suite.app, suite.ctx            // get ctx after init
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:180]: nit: please avoid redundant comments 👍 
```suggestion
	lenderAddr, _ := suite.initBorrowScenario()
	app, ctx := suite.app, suite.ctx
```

File: x/leverage/keeper/keeper_test.go [hunk 9]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
...
+
+	// The "lender" user from the init scenario is being used because it
+	// already has 1k u/umee for collateral
+
+	// lender attempts to borrow 200 u/umee, fails because uTokens cannot be borrowed
+	err := app.LeverageKeeper.BorrowAsset(ctx,
+		lenderAddr,
+		tCoin("u/umee", 200),
+	)
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:189]: nit: this can be a single line IMO

File: x/leverage/keeper/keeper_test.go [hunk 10]
```diff
@@ -128,6 +147,314 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
...
+	suite.Require().Error(err)
+}
+
+func (suite *IntegrationTestSuite) TestGetCollateral_Invalid() {
+	lenderAddr, _ := suite.initBorrowScenario() // create initial conditions
+	app, ctx := suite.app, suite.ctx            // get ctx after init
+
+	// Any user from the starting scenario can be used, since we are only viewing
+	// collateral settings.
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper_test.go:441]: nit: this applies to all the tests. Try to keep things on one line if they can fit under 120 chars. Also, avoid redundant comments.

**Review by alexanderbez — CHANGES_REQUESTED** [review]
@toteki this looks great! Nice work. Left some comments, mainly about readability 👍

File: x/leverage/keeper/keeper.go [hunk 11]
```diff
@@ -123,3 +124,175 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
+	}
+	// Determine the total amount of denom borrowed (previously borrowed + newly borrowed)
+	totalBorrowed := currentlyBorrowed.AmountOf(loan.Denom).Add(loan.Amount)
+	// Store the new borrowed amount in keeper
+	store := ctx.KVStore(k.storeKey)
+	b, err := totalBorrowed.Marshal()
+	if err != nil {
+		return err
+	}
+	store.Set(types.CreateLoanKey(borrowerAddr, loan.Denom), b)
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go:245]: What has been marshaled into bytes in the previous few lines is an sdk.Int representing the desired denom, so we're in line with what you expect.

File: x/leverage/keeper/keeper_test.go [hunk 12]
```diff
@@ -128,6 +147,279 @@ func (suite *IntegrationTestSuite) TestWithdrawAsset_Valid() {
...
+	suite.Require().Equal(tokenBalance, tCoin("umee", 9000))
+
+	// verify lender's uToken balance remains at 1k from initial conditions
+	uTokenBalance := app.BankKeeper.GetBalance(ctx, lenderAddr, "u/"+umeeapp.BondDenom)
+	suite.Require().Equal(uTokenBalance, tCoin("u/umee", 1000))
+
+	// lender repays 50 umee - this time it fails because the loan no longer exists
+	err = app.LeverageKeeper.RepayAsset(ctx, lenderAddr, tCoin("umee", 50))
+	suite.Require().Error(err)
+}
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:341]: Would it be okay to keep some redundant comments inside the _test files, when they're inside functions where the majority of blocks have meaningful comments?

For example in the highlighted function, I know `// lender borrows 200 umee` has no need to be a comment, but its presence makes the test function itself very stylistically consistent (each meaningful action is a whitespace-separated code block with a leading comment, followed by actions then additional checks, if any).

File: x/leverage/keeper/keeper_test.go [hunk 13]
```diff
@@ -23,6 +24,24 @@ var (
 	initCoins  = sdk.NewCoins(sdk.NewCoin(umeeapp.BondDenom, initTokens))
 )
 
+// makes sdk.Coin for testing (easier to read)
+func tCoin(denom string, amount int) sdk.Coin {
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper_test.go:28]: It's strictly for readability.

Besides hitting the 120 character line limit in some of the earlier iterations of my require statements, I noticed that there were lines in existing tests where it took some squinting to tell the difference between 1k and 10k, e.g.
`err := app.LeverageKeeper.LendAsset(ctx, lenderAddr, sdk.NewInt64Coin(umeeapp.BondDenom, 1000000000)) // 1k umee`

tCoin adds the 6 zeroes and does the required transform on the denom (e.g. `1 "umee" => 1000000 "uumee"`) to make amounts easier to follow at a glance.

So `sdk.NewInt64Coin(umeeapp.BondDenom, 1000000000) //1k umee`
becomes `tCoin("umee",1000)`

Might have been overkill; let me know if you think I should remove it after that.

[the remark this answers is not in the document]
Comment (alexanderbez) [reply x/leverage/keeper/keeper_test.go:28]: Yeah I think this is totally overkill. Also, the function does denom checking with `/u` stuff which I don't really follow.

Instead, consider keeping `sdk.NewInt64Coin(umeeapp.BondDenom, ...)` calls, but instead store the amount in a global private `var`.

File: x/leverage/keeper/keeper.go [hunk 14]
```diff
@@ -123,3 +124,163 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
+		return false
+	}
+	store := ctx.KVStore(k.storeKey)
+	// Any value (expected = 0x01) found at key will be interpreted as true.
+	key := types.CreateCollateralSettingKey(borrowerAddr, denom)
+	return store.Has(key)
+}
+
+// GetLoan returns an sdk.Coin representing how much of a given denom a borrower currently owes.
+func (k Keeper) GetLoan(ctx sdk.Context, borrowerAddr sdk.AccAddress, denom string) sdk.Coin {
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go:157]: This function returns an sdk.Coin, but what was stored in keeper and unmarshaled from bytes was just the amount (an sdk.Int). See the line
`err := owed.Amount.Unmarshal(bz)`
where the sdk.Int's unmarshal method is used.

I could simplify the function to return just the Int and not construct a coin around it though - should I do that?

[the remark this answers is not in the document]
Comment (alexanderbez) [reply x/leverage/keeper/keeper.go:157]: ahh you're totally right, sorry I missed that! No I think this is fine.

File: x/leverage/keeper/keeper.go [hunk 15]
```diff
@@ -123,3 +123,161 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
+	store := ctx.KVStore(k.storeKey)
+	owed := sdk.NewCoin(denom, sdk.ZeroInt())
+	key := types.CreateLoanKey(borrowerAddr, denom)
+	bz := store.Get(key)
+	if bz != nil {
+		err := owed.Amount.Unmarshal(bz)
+		if err != nil {
+			// Should never happen
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:164]:
```suggestion
```
